### PR TITLE
Add LanguageIdentifier iai benches

### PR DESCRIPTION
**Pull request by zbraniecki**
Adding first IAI benchmarks. I kept it minimal to get over the hump of how to add iai benches, and then I hope to get help spreading iai benchmarks to other components.

The limitation is that iai works only with valgrind and valgrind currently is not available on M1 archs and I doubt we'll have it on Windows. That leaves intel macs and linux.

I worked it around on my m1 MBA by installing Ubuntu `multipass`.

The results are:

```
bench_langid_constr
  Instructions:                2659 (No change)
  L1 Accesses:                 3348 (No change)
  L2 Accesses:                    2 (No change)
  RAM Accesses:                  47 (No change)
  Estimated Cycles:            5003 (No change)

bench_langid_compare_components
  Instructions:                   3 (No change)
  L1 Accesses:                    2 (No change)
  L2 Accesses:                    2 (No change)
  RAM Accesses:                   0 (No change)
  Estimated Cycles:              12 (No change)

bench_langid_compare_components_str
  Instructions:                  10 (No change)
  L1 Accesses:                   12 (No change)
  L2 Accesses:                    2 (No change)
  RAM Accesses:                   0 (No change)
  Estimated Cycles:              22 (No change)

bench_langid_matching
  Instructions:                   3 (No change)
  L1 Accesses:                    2 (No change)
  L2 Accesses:                    2 (No change)
  RAM Accesses:                   0 (No change)
  Estimated Cycles:              12 (No change)

bench_langid_matching_str
  Instructions:                1260 (No change)
  L1 Accesses:                 1412 (No change)
  L2 Accesses:                    3 (No change)
  RAM Accesses:                  23 (No change)
  Estimated Cycles:            2232 (No change)

bench_langid_serialize
  Instructions:                4588 (No change)
  L1 Accesses:                 6656 (No change)
  L2 Accesses:                   11 (No change)
  RAM Accesses:                  42 (No change)
  Estimated Cycles:            8181 (No change)

bench_langid_canonicalize
  Instructions:                6964 (No change)
  L1 Accesses:                 9694 (No change)
  L2 Accesses:                   14 (No change)
  RAM Accesses:                  67 (No change)
  Estimated Cycles:           12109 (No change)
```

### CODE REVIEW

**Comment by zbraniecki** [comment]
I consider any CI around those tests out of scope for this PR. That means that we won't get nice alerts when we regress and it's up to a patch author to test against those.

I am open to introduce a new feature for IAI tests like "iai_bench" to make DX of testing just IAI easier. I'll let reviewers decide what they prefer. We can also punt this question to after we have a bunch of IAI benchmarks and get a chance to observe the experience of working with them.

File: components/locid/benches/iai_langid.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+// This file is part of ICU4X. For terms of use, please see the file
```
Comment (sffc) [inline components/locid/benches/iai_langid.rs:1]: Request: Add one for `cmp_bytes` since this is (or will be) the hot path for data loading.

**Comment by nciric** [comment]
I had to read about it at https://bheisler.github.io/criterion.rs/book/iai/getting_started.html.

It seems it's still work in progress, missing some CI features etc. I may have missed it, but what's the rush to integrate with iai vs Criterion we already have? (I assume we'll keep both)

**Comment by zbraniecki** [comment]
> It seems it's still work in progress, missing some CI features etc. I may have missed it, but what's the rush to integrate with iai vs Criterion we already have? (I assume we'll keep both)

Iai is a "wip" for a couple years now, mostly in the sense of the author (it's by the same person as criterion) figuring out how to merge both.
The value proposition is that `iai` gives us CPU independent, no-noise, high-signal information allowing us to test impact of any change on the instruction count.

It pays off really well in certain scenarios which I'm about to evaluate such as CPT optimizations and other low-level algorithmic adjustments.
I'm placing all of the new benchmarks behind `bench` feature, so they won't impact anyone's day-2-day work.

It also places us on the path to fix #397 which is something Shane has been advocating for since the beginning of the project.

To summarize - I think there's a value for people working on low-level perf optimizations and no cost of adding it now.

**Review by nciric — APPROVED** [review]

**Review by sffc — APPROVED** [review]
